### ISSUE TITLE
Warning 302: Identifier '<named union in struct>' redefined (ignored)

### ISSUE THREAD

**Issue by kenr** [swig/swig#1305]
Warning 302 is issued and no code is generated for a struct member of type union using a name used previously. 

Here is a swig example reproducing the issue:
```
%module "test"

%inline %{
    typedef struct {
        union x_union {
            int a;
        } duplicate_p;
    } x_t;

    typedef struct {
        union y_union {
            int b;
        } duplicate_p;
    } y_t;
%}

```

Running swig -python <testfile> reproduces the issue.

Removing the union name (x_union, y_union) works, but the C API we wrap with SWIG uses the C syntax union <union_name> {<union types>} <struct member_name>.

I've tried to circumvent this issue with %rename, with no luck.

Swig version info:
SWIG Version 3.0.12
Compiled with i686-w64-mingw32-g++ [i686-w64-mingw32]
Configured options: +pcre

**Comment by wsfulton** [comment]
This seems to work:
```swig
%rename(x_duplicate_p) x_union;
%rename(y_duplicate_p) y_union;

%inline %{
    typedef struct X {
        struct x_union {
            int a;
        } duplicate_p;
    } x_t;

    typedef struct Y {
        struct y_union {
            int b;
        } duplicate_p;
    } y_t;
%}
```
but emits a spurious warning:
```
example.i:32: Warning 302: Identifier 'duplicate_p' redefined (ignored),
example.i:26: Warning 302: previous definition of 'duplicate_p'.
```
So that needs fixing.

Alternatively use:

```swig
%inline %{
#ifdef SWIG
%rename(x_duplicate_p) duplicate_p;
#endif
    typedef struct {
        union x_union {
            int a;
        } duplicate_p;
    } x_t;

#ifdef SWIG
%rename(y_duplicate_p) duplicate_p;
#endif
    typedef struct {
        union y_union {
            int b;
        } duplicate_p;
    } y_t;
%}
```
or put the unions in different .h files and then add the %rename before each of the `%include` of each .h file.

@wkalinin any other ideas?

**Comment by kenr** [comment]
@wsfulton Thank you for a quick reply.

I tried the first suggestion, but there is no duplicate_p object attribute added to y_t. x_t has the duplicate_p object attribute. Perhaps I'm using the wrong version of SWIG (3.0.12)? 

```
class x_t(_object):
    __swig_setmethods__ = {}
    __setattr__ = lambda self, name, value: _swig_setattr(self, x_t, name, value)
    __swig_getmethods__ = {}
    __getattr__ = lambda self, name: _swig_getattr(self, x_t, name)
    __repr__ = _swig_repr
    __swig_setmethods__["duplicate_p"] = _test.x_t_duplicate_p_set
    __swig_getmethods__["duplicate_p"] = _test.x_t_duplicate_p_get
    if _newclass:
        duplicate_p = _swig_property(_test.x_t_duplicate_p_get, _test.x_t_duplicate_p_set)

    def __init__(self):
        this = _test.new_x_t()
        try:
            self.this.append(this)
        except __builtin__.Exception:
            self.this = this
    __swig_destroy__ = _test.delete_x_t
    __del__ = lambda self: None
x_t_swigregister = _test.x_t_swigregister
x_t_swigregister(x_t)

class y_t(_object):
    __swig_setmethods__ = {}
    __setattr__ = lambda self, name, value: _swig_setattr(self, y_t, name, value)
    __swig_getmethods__ = {}
    __getattr__ = lambda self, name: _swig_getattr(self, y_t, name)
    __repr__ = _swig_repr

    def __init__(self):
        this = _test.new_y_t()
        try:
            self.this.append(this)
        except __builtin__.Exception:
            self.this = this
    __swig_destroy__ = _test.delete_y_t
    __del__ = lambda self: None
y_t_swigregister = _test.y_t_swigregister
y_t_swigregister(y_t)

```
The second suggestion works, but I do not have the flexibility to change the existing headers.

The third suggestion would work if I make a copy of the existing headers and modify them as suggested. 
This adds some extra maintenance though.

**Comment by kenr** [comment]
I found another workaround by using the duplicate detection as a feature by using SWIG inline before includes. That way I can remove the union name (union <remove_union_name> { <structs> } params) in the inlined code.

Here is an example:
```
%module "test"

%inline %{
    // SWIG inline
    typedef struct fwd_decl_struct_t fwd_decl_struct_t;

    typedef struct {
        union {
            int a;
        } duplicate_p;
    } x_t;

    typedef struct {
        union {
            fwd_decl_struct_t b;
        } duplicate_p;
    } y_t;

    // Simulated .h file
    typedef struct {
        union x_union {
            int a;
        } duplicate_p;
    } x_t;

    typedef struct
    {
        int inside_fwd_decl_struct;
    } fwd_decl_struct_t;

    typedef struct {
        union y_union {
            fwd_decl_struct_t b;
        } duplicate_p;
    } y_t;
%}
```

It is a bit of work and double maintenance since the project I swig is quite large so if SWIG could support the "union <union_name> { <structs> } <name>" syntax it would be great.

Let me know if I can assist.

**Comment by wkalinin** [comment]
This looks like a bug with nested structures symbol tables in "C" mode. I think it can be fixed by setting correct namespace: https://github.com/swig/swig/pull/1306

**Comment by wsfulton** [comment]
I've merged in #1306 to fix. Thanks @wkalinin. @kenr, please check.

**Comment by wkalinin** [comment]
Thanks for the test William, I've only Windows box at hand, and running the test suite here is not quite trivial.
